ephios/plugins/basesignup/signup/section_based.py

```diff
@@ -0,0 +1,319 @@
+                    preferred_section_uuid=form.cleaned_data["section"],
+                )
+                messages.success(
+                    self.request,
+                    self.method.signup_success_message.format(shift=self.shift),
+                )
+        except ParticipationError as errors:
+            for error in errors:
+                messages.error(self.request, self.method.signup_error_message.format(error=error))
+        return redirect(self.shift.event.get_absolute_url())
+
+    def signup_pressed(self, request, *args, **kwargs):
+        if not self.method.configuration.choose_preferred_section:
+            return super().signup_pressed(request, *args, **kwargs)
+        return redirect(reverse("event_management:shift_action", kwargs=dict(pk=self.shift.pk)))
+
+
+class SectionBasedSignupMethod(BaseSignupMethod):
+    slug = "section_based"
+    verbose_name = _("Apply for sections")
+    description = _(
+        """This method lets you define sections for which people can choose from.
+        Sections contain qualifications that helpers need to fulfil."""
+    )
+    registration_button_text = _("Request")
+    signup_success_message = _("You have successfully requested a participation for {shift}.")
+    signup_error_message = _("Requesting a participation failed: {error}")
+    configuration_form_class = SectionBasedConfigurationForm
+    signup_view_class = SectionBasedSignupView
+
+    def get_configuration_fields(self):
+        return {
+            **super().get_configuration_fields(),
+            "choose_preferred_section": {
+                "formfield": forms.BooleanField(
+                    label=_("Ask participants for a preferred section"),
+                    help_text=_("This only makes sense if you configure multiple sections."),
+                    widget=forms.CheckboxInput,
+                    required=False,
+                ),
+                "default": False,
+            },
+            "sections": {
+                "formfield": forms.Field(
+                    label=_("Structure"),
+                    widget=forms.HiddenInput,
+                    required=False,
+                ),
+                "default": [],
+            },
+        }
+
+    @staticmethod
+    def check_qualification(method, participant):
+        if not sections_participant_qualifies_for(method.configuration.sections, participant):
+            return ParticipationError(_("You are not qualified."))
+
+    @property
+    def signup_checkers(self):
+        return super().signup_checkers + [self.check_qualification]
+
+    def perform_signup(self, participant: AbstractParticipant, preferred_section_uuid=None):
+        participation = super().perform_signup(participant)
+        participation.data["preferred_section_uuid"] = preferred_section_uuid
+        if preferred_section_uuid:
+            # reset dispatch decision, as that would have overwritten the preferred choice
```

Comment:
in what scenario would we want that reset?

Reply:
If you register and get dispatched, then get declined and later request participation again.

Reply:
> Scenario: A participation has been confirmed (and therefore a section has been selected) and the user declines the participation afterwards. Once the user requests a participation again the request is displayed with the section he once had even if users are not allowed to pick a section.

(your comment from elsewhere)
This is intended behavior IMO?!